### DISCUSSION TITLE
Rstudio IDE and Snippets

### DISCUSSION THREAD

**Discussion by rressler** [quarto-dev/quarto-cli#1944] [Q&A]
I can't see a way to use quarto snippets, e.g. for panel-tabset, in the RStudio IDE outside a code chunk. Is there a way or is this a future feature?

**Comment by mcanouil** [comment]
Just appeared on my Twitter feed: <https://gist.github.com/jthomasmock/11acebd4448f171f786e01397df34116> by @jthomasmock

Although, I wonder what you mean by "in the RStudio IDE outside a code chunk" ... 🤔
Snippets are not inside or outside anything besides the active document/file.

To Quarto maintainers, maybe worth adding the snippets to the documentation? (or to the RStudio Add-ins if not already there and in the VSCode extension)

**Reply by mcanouil** [reply]
To complete my answer and as highlighted by @andrewpbray choose markdown as the scope for the snippets, not R (the default). A snippet completion is triggered by `shift + tab` (at least on Windows)
![image](https://user-images.githubusercontent.com/8896044/185270061-414c7dea-25e7-468b-9f53-394917978d65.png)

The same principle works for VSCode, the language/document type has to bet set (this is described by @pat-s on the gist).
For RStudio IDE, you can take a look at @jjallaire post https://support.rstudio.com/hc/en-us/articles/204463668-Code-Snippets-in-the-RStudio-IDE
In it, you'll see the file location (if you want to edit it outside of RStudio):
- On Windows `%appdata%/Roaming/RStudio/snippets/`
- On Unix-based OS `~/.config/rstudio/snippets/`

![image](https://user-images.githubusercontent.com/8896044/185270921-40cf428f-2824-4f05-a4bf-dafd657d89f1.png)

Any snippets within `markdown.snippets` will work for files with `rmd`, `rmarkdown`, and `qmd` extensions. For some reasons, the snippets are unavailable in `md` and `markdown` file 🤷‍♂️ (see <https://github.com/rstudio/rstudio/issues/11777>)

**Comment by rressler** [comment]
Mickael, thanks - Yes,
But there is no way to invoke these inside the RStudio visual or source editor that I can see without being in a code chunk that you then have to delete.
That is the question. The documentation talks about doing it inside the VS Code editor.\Thanks,
Richard


From: Mickaël Canouil ***@***.***>
Reply-To: quarto-dev/quarto-cli ***@***.***>
Date: Wednesday, August 17, 2022 at 10:59 AM
To: quarto-dev/quarto-cli ***@***.***>
Cc: Richard Ressler ***@***.***>, Author ***@***.***>
Subject: Re: [quarto-dev/quarto-cli] Rstudio IDE and Snippets (Discussion #1944)


External Email: Use caution with links and attachments.

Just appeared on my Twitter feed: https://gist.github.com/jthomasmock/11acebd4448f171f786e01397df34116?permalink_comment_id=4269591#gistcomment-4269591<https://urldefense.com/v3/__https:/gist.github.com/jthomasmock/11acebd4448f171f786e01397df34116?permalink_comment_id=4269591*gistcomment-4269591__;Iw!!IaT_gp1N!0x4kXH6UltWsnxrPsvYHEFWXfviWcn2t0eE92kBWId6wyu2GQaKH-Kpl1pW-E5yv8gVOqDb5FZ3mTZ2Tr4chKalHCw0$> by @jthomasmock<https://urldefense.com/v3/__https:/github.com/jthomasmock__;!!IaT_gp1N!0x4kXH6UltWsnxrPsvYHEFWXfviWcn2t0eE92kBWId6wyu2GQaKH-Kpl1pW-E5yv8gVOqDb5FZ3mTZ2Tr4chTdVN4dc$>

**Reply by mcanouil** [reply]
I don't know about the visual editor but last time I Checked I never had to be in a code chunk for a snippets to work.

**Reply by jthomasmock** [reply]
To get the snippet to "replace" inside markdown you have to use the <kbd>shift + tab</kbd> shortcut rather than just the <kbd>tab</kbd> shortcut that you're likely used to for auto-complete of code.

https://user-images.githubusercontent.com/29187501/185264603-4a80d966-851f-4352-a769-cb836628335c.mp4

<hr>

Also note that as @andrewpbray indicated - you'll need to create new snippets from inside the Markdown snippets section rather than R code section.

**Comment by andrewpbray** [comment]
The first time I tried copying in snippets to make quarto fenced divs and bracketed spans, I copied them into the first file that came up in the RStudio > Tools > Edit Code Snippets. It didn't work, then I realized that that first file is for snippets inside R code cells. For snippets inside markdown, I had to toggle to that tab on the left; that worked.

<img width="571" alt="Screen Shot 2022-08-17 at 3 40 58 PM" src="https://user-images.githubusercontent.com/1627362/185256316-011cc8d5-36bb-48f2-a6a6-1def77480e20.png">

I'm not sure if this is the same problem that you're running into?

**Reply by mcanouil** [reply]
Yep, the scope matters.
I completed my initial answer with more details (see <https://github.com/quarto-dev/quarto-cli/discussions/1944#discussioncomment-3415648>).

**Comment by rressler** [comment]
Mickaël and Tom,
Thanks for the helpful and quick answers. I am using a Mac and once the custom snippets are placed under R Markdown instead of R (my bad), the snippets work fine with Shift Tab in the source editor but there is no auto-complete. You have to type the whole snippet label then hit Shift Tab.

The RStudio Visual editor includes “Tabset” and “Calllout” under the Insert  and CMD / options. It might be nice to have a Snippet option that allows you to select a custom one based on its label.

Thanks so much!
Richard

From: Mickaël Canouil ***@***.***>
Reply-To: quarto-dev/quarto-cli ***@***.***>
Date: Wednesday, August 17, 2022 at 9:17 PM
To: quarto-dev/quarto-cli ***@***.***>
Cc: Richard Ressler ***@***.***>, Author ***@***.***>
Subject: Re: [quarto-dev/quarto-cli] Rstudio IDE and Snippets (Discussion #1944)


External Email: Use caution with links and attachments.

To complete my answer and as highlighted by @andrewpbray<https://urldefense.com/v3/__https:/github.com/andrewpbray__;!!IaT_gp1N!0wb3D1kSA02za0q0oSzRICPLaqZGQqVvlRNxmQAtnhHeVO2Gla9v4yDK3YXY7QESjb2au3RvZGEvh9dffEKktrO1TCc$> choose markdown as the scope for the snippets, not R (the default). A snippet completion is triggered by shift + tab (at least on Windows)
[Image removed by sender. image]<https://urldefense.com/v3/__https:/user-images.githubusercontent.com/8896044/185270061-414c7dea-25e7-468b-9f53-394917978d65.png__;!!IaT_gp1N!0wb3D1kSA02za0q0oSzRICPLaqZGQqVvlRNxmQAtnhHeVO2Gla9v4yDK3YXY7QESjb2au3RvZGEvh9dffEKkGPVH-II$>

The same principle works for VSCode, the language/document type has to bet set (this is described by @pat-s<https://urldefense.com/v3/__https:/github.com/pat-s__;!!IaT_gp1N!0wb3D1kSA02za0q0oSzRICPLaqZGQqVvlRNxmQAtnhHeVO2Gla9v4yDK3YXY7QESjb2au3RvZGEvh9dffEKk2Gbq0X0$> on the gist).
For RStudio IDE, you can take a look at @jjallaire<https://urldefense.com/v3/__https:/github.com/jjallaire__;!!IaT_gp1N!0wb3D1kSA02za0q0oSzRICPLaqZGQqVvlRNxmQAtnhHeVO2Gla9v4yDK3YXY7QESjb2au3RvZGEvh9dffEKk52vHAnQ$> post https://support.rstudio.com/hc/en-us/articles/204463668-Code-Snippets-in-the-RStudio-IDE<https://urldefense.com/v3/__https:/support.rstudio.com/hc/en-us/articles/204463668-Code-Snippets-in-the-RStudio-IDE__;!!IaT_gp1N!0wb3D1kSA02za0q0oSzRICPLaqZGQqVvlRNxmQAtnhHeVO2Gla9v4yDK3YXY7QESjb2au3RvZGEvh9dffEKk2Ugbl0M$>
In it, you'll see the file location (if you want to edit it outside of RStudio):

  *   On Windows %appdata%/Roaming/RStudio/snippets/
  *   On Unix-based OS ~/.config/rstudio/snippets/

**Reply by mcanouil** [reply]
RStudio IDE does not autocomplete/suggest for snippets on Windows and on Mac apparently while VSCode does.
Although, all this are in the end not a Quarto matter, but an RStudio IDE matter.
Maybe worth asking for enhancement in https://github.com/rstudio/rstudio and set https://github.com/quarto-dev/quarto-cli/discussions/1944#discussioncomment-3415648 as answer since it answers your main question.
